research/cifar10/adaptive_pfl/ditto/server.py

```diff
@@ -0,0 +1,123 @@
+    n_server_rounds: int,
+    n_clients: int,
+    current_server_round: int,
+) -> Config:
+    return {
+        "batch_size": batch_size,
+        "local_epochs": local_epochs,
+        "n_server_rounds": n_server_rounds,
+        "n_clients": n_clients,
+        "current_server_round": current_server_round,
+    }
+
+
+def main(config: Dict[str, Any], server_address: str, lam: float, adapt_loss_weight: bool) -> None:
+    # This function will be used to produce a config that is sent to each client to initialize their own environment
+    fit_config_fn = partial(
+        fit_config,
+        config["batch_size"],
+        config["local_epochs"],
+        config["n_server_rounds"],
+        config["n_clients"],
+    )
+
+    client_manager = SimpleClientManager()
+    # Initializing the model on the server side
+    model = ConvNet(in_channels=3, use_bn=False, dropout=0.1)
+    # Server performs simple FedAveraging as its server-side optimization strategy
+    strategy = FedAvgWithAdaptiveConstraint(
+        min_fit_clients=config["n_clients"],
+        min_evaluate_clients=config["n_clients"],
+        # Server waits for min_available_clients before starting FL rounds
+        min_available_clients=config["n_clients"],
+        on_fit_config_fn=fit_config_fn,
+        # We use the same fit config function, as nothing changes for eval
+        on_evaluate_config_fn=fit_config_fn,
+        fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+        evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+        initial_parameters=get_all_model_parameters(model),
+        initial_loss_weight=lam,
+        adapt_loss_weight=adapt_loss_weight,
+    )
+
+    server = PersonalServer(client_manager=client_manager, strategy=strategy)
+
+    fl.server.start_server(
+        server=server,
+        server_address=server_address,
+        config=fl.server.ServerConfig(num_rounds=config["n_server_rounds"]),
+    )
+
+    log(INFO, "Training Complete")
+    log(INFO, f"Best Aggregated (Weighted) Loss seen by the Server: \n{server.best_aggregated_loss}")
+
+    # Shutdown the server gracefully
+    server.shutdown()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="FL Server Main")
+    parser.add_argument(
+        "--config_path",
+        action="store",
+        type=str,
+        help="Path to configuration file.",
+        default="config.yaml",
+    )
+    parser.add_argument(
+        "--server_address",
+        action="store",
+        type=str,
+        help="Server Address to be used to communicate with the clients",
+        default="0.0.0.0:8080",
+    )
+    parser.add_argument(
+        "--seed",
+        action="store",
+        type=int,
+        help="Seed for the random number generators across python, torch, and numpy",
+        required=False,
+    )
+    parser.add_argument(
+        "--lam", action="store", type=float, help="Ditto loss weight for local model training", default=0.01
```

Comment:
Can we spell out `lambda` instead? Just for more clarity. Here and also on the variables and parameters.

Reply:
Will do.